tests/test_common_gpu.py

```diff
@@ -356,3 +357,45 @@ def test_modules_to_save_grad(self):
         self.assertTrue(modules_to_save.weight.requires_grad is True)
         self.assertTrue(original_module.weight.grad is None)
         self.assertTrue(modules_to_save.weight.grad is not None)
+
+    @require_torch_gpu
+    @pytest.mark.single_gpu_tests
+    @require_bitsandbytes
+    def test_8bit_merge_lora(self):
+        model = AutoModelForCausalLM.from_pretrained(
+            "facebook/opt-125m",
+            load_in_8bit=True,
+        )
+        config = LoraConfig(
+            r=8,
+        )
+        model = get_peft_model(model, config)
+        model.merge_and_unload("default")
+        self.assertTrue(isinstance(model, PeftModel))
+        self.assertTrue(
+            isinstance(model.base_model.model.model.decoder.layers[0].self_attn.q_proj, bnb.nn.Linear8bitLt)
+        )
+        self.assertTrue(
```

Comment:
Can you add a logits or generation test for that ? you can check the logits of the model before and after merging and make sure they are the same. Make sure to add `init_lora_weights=False` when creating the LoraConfig to avoid initializing the B matrix with full zeros: https://github.com/huggingface/peft/blob/main/src/peft/tuners/lora.py#L98

Reply:
Hi @younesbelkada 
It is hard to keep the result of forward logits or generation text the same because both dequantization and quantization have precision loss. You can try the following script
```python
import torch
from transformers import AutoModelForCausalLM, BitsAndBytesConfig

torch.manual_seed(1024)
model_origin = AutoModelForCausalLM.from_pretrained(
                "opt-125m",
                torch_dtype=torch.float32,
            )
bnb_config = BitsAndBytesConfig(
            load_in_4bit=True,
            bnb_4bit_use_double_quant=True,
            bnb_4bit_compute_type=torch.float32
        )
model = AutoModelForCausalLM.from_pretrained(
            "opt-125m",
            quantization_config=bnb_config,
            torch_dtype=torch.float32,
        )
random_input = torch.LongTensor([[1, 0, 1, 0, 1, 0]]).to(model.device)
print(model_origin(random_input.clone().to(model_origin.device)).logits)
print(model(random_input).logits)
```

The output should be 
```
tensor([[[-3.9464, -3.9443,  3.2428,  ..., -3.9583, -3.9531, -4.0695],
         [ 2.1592,  2.1657,  3.4937,  ...,  2.1609,  2.2064,  1.8032],
         [ 1.7407,  1.7366,  3.3675,  ...,  1.7371,  1.7456,  1.5533],
         [ 1.7655,  1.7583,  3.1950,  ...,  1.7668,  1.7512,  1.5889],
         [ 1.8861,  1.8778,  3.0723,  ...,  1.8935,  1.8632,  1.7100],
         [ 2.0078,  1.9973,  2.9639,  ...,  2.0184,  1.9726,  1.8311]]],
       grad_fn=<UnsafeViewBackward0>)
tensor([[[-5.2579, -5.2552,  2.2640,  ..., -5.2631, -5.2835, -5.2650],
         [ 1.0482,  1.0549,  3.4282,  ...,  1.0684,  1.1276,  0.7290],
         [ 0.8263,  0.8238,  3.5503,  ...,  0.8330,  0.8313,  0.5981],
         [ 0.8692,  0.8614,  3.1840,  ...,  0.8806,  0.8476,  0.6761],
         [ 0.9691,  0.9599,  3.1258,  ...,  0.9846,  0.9382,  0.7884],
         [ 1.0963,  1.0847,  3.1242,  ...,  1.1106,  1.0566,  0.9201]]],
       device='cuda:0', grad_fn=<UnsafeViewBackward0>)
```

So the quantization error is large. Since we dequantize the weight, and then add lora weight, and quantize the new weight, the loss error will be too large.

Reply:
I see, thanks for explaining, I have quickly played with your script. I think that despite the logits are slightly different the argmax should stay the same, I believe this would be a sufficient test. 

```python
import torch
from transformers import AutoModelForCausalLM, BitsAndBytesConfig

torch.manual_seed(1024)
model_origin = AutoModelForCausalLM.from_pretrained(
    "facebook/opt-125m",
    torch_dtype=torch.float32,
)
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_use_double_quant=False,
    bnb_4bit_compute_type=torch.float32
)
model = AutoModelForCausalLM.from_pretrained(
    "facebook/opt-125m",
    quantization_config=bnb_config,
    torch_dtype=torch.float32,
)
random_input = torch.LongTensor([[1, 0, 1, 0, 1, 0]]).to(model.device)

print(model_origin.generate(random_input.clone().to(model_origin.device), max_new_tokens=1))
print(model.generate(random_input.clone().to(model_origin.device), max_new_tokens=1))
```

This should give:

```bash
tensor([[    1,     0,     1,     0,     1,     0, 50118]])
tensor([[    1,     0,     1,     0,     1,     0, 50118]])
```

Could you add a test that tests that the first generated token is the same between the merged and non-merged model? 🙏 Thanks again for your work on this!

Reply:
I ran the 8bit test locally on my GPU and see large differences in the logits (`l0` are original logits, `l1` merged):

```python
>>> torch.testing.assert_close(l0, l1)
*** AssertionError: Tensor-likes are not close!
Mismatched elements: 301245 / 301632 (99.9%)
Greatest absolute difference: 9.671875 at index (0, 5, 43954) (up to 1e-05 allowed)
Greatest relative difference: 29409.54607899256 at index (0, 2, 1238) (up to 0.001 allowed)

>>> torch.abs(l0-l1).mean() / torch.abs(l0).mean()
tensor(0.6348, device='cuda:0', dtype=torch.float16)

>>> l0.mean()
tensor(-2.6367, device='cuda:0', dtype=torch.float16)
>>> l1.mean()
tensor(-3.7637, device='cuda:0', dtype=torch.float16)
```

Also plotted a sample of logits against another:

![Screenshot from 2023-08-25 13-54-44](https://github.com/huggingface/peft/assets/6229650/a214f02e-c2f8-4be9-a9be-2f95c0f82755)

In contrast, this is what I find for 4bit:

```python
>>> torch.testing.assert_close(l0, l1)
*** AssertionError: Tensor-likes are not close!

Mismatched elements: 301627 / 301632 (100.0%)
Greatest absolute difference: 1.9874346256256104 at index (0, 4, 5985) (up to 1e-05 allowed)
Greatest relative difference: 4337.804713804714 at index (0, 3, 3491) (up to 1.3e-06 allowed)

>>> torch.abs(l0-l1).mean() / torch.abs(l0).mean()
tensor(0.1003, device='cuda:0')
>>> l0.mean()
tensor(-2.9331, device='cuda:0')
>>> l1.mean()
tensor(-3.0172, device='cuda:0')
```

And the scatter plot:

![Screenshot from 2023-08-25 14-00-38](https://github.com/huggingface/peft/assets/6229650/a29abbeb-f4bc-4aaa-b624-d4ce9422b822)

The differences can also be quite big but it's much better than 8bit. Most notably, 8bit has much higher deviation _and_ bias.

As is, I think I wouldn't feel comfortable exposing 8bit merging to users, whereas 4bit looks fine.

Reply:
Small update, I changed the code to set the lora weights to 0 and still found the deviations to be of the same magnitude. So it looks like it's just the back and forth conversion that introduces the error, not the addition of lora weights.

Reply:
Great insights Benjamin!